pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,455 @@
+	c.APIKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.APIKey, "PACKET_API_KEY")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"apiKey\" field, error = %v", err)
+	}
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"instanceType\" field, error = %v", err)
+	}
+	c.BillingCycle, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.BillingCycle)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"billingCycle\" field, error = %v", err)
+	}
+	if c.BillingCycle == "" {
+		c.BillingCycle = defaultBillingCycle
+	}
+	for i, tag := range rawConfig.Tags {
+		tagValue, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get the %d value of \"tags\" field, error = %v", i, err)
+		}
+		c.Tags = append(c.Tags, tagValue)
+	}
+	for i, facility := range rawConfig.Facilities {
+		facilityValue, err := p.configVarResolver.GetConfigVarStringValue(facility)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get the %d value of \"facilities\" field, error = %v", i, err)
+		}
+		c.Facilities = append(c.Facilities, facilityValue)
+	}
+	return &c, &rawConfig, &pconfig, err
+}
+
+func (p *provider) getPacketDevice(machine *v1alpha1.Machine) (*packngo.Device, *packngo.Client, error) {
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, generateTag(string(machine.UID)))
+	if err != nil {
+		return nil, nil, err
+	}
+	return device, client, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, _, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.APIKey == "" {
+		return errors.New("apiKey is missing")
+	}
+	if c.InstanceType == "" {
+		return errors.New("instanceType is missing")
+	}
+	if c.ProjectID == "" {
+		return errors.New("projectID is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.APIKey)
+
+	if len(c.Facilities) == 0 || c.Facilities[0] == "" {
+		return fmt.Errorf("must have at least one non-blank facility")
+	}
+
+	// get all valid facilities
+	facilities, _, err := client.Facilities.List(nil)
```

Comment:
Not really something to do with the PR, but the `response Body`/`Request URL`/`Curl` fields in your API docs about this are pretty empty: https://www.packet.com/developers/api/#retrieve-all-facilities

Reply:
Yes, I agree 100%. Going to post internally about this (and the `userdata` issue we saw earlier).